doc/nom_recipes.md

```diff
@@ -0,0 +1,235 @@
+  )(i)
+}
+```
+
+### `/* C-style comments */`
+
+Inline comments surrounded with sentinel tags `(*` and `*)`. This version returns an output of `()` and does not handle nested comments.
+
+```rust
+pub fn pinline_comment<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, (), E> {
+  map(
+    tuple((
+      tag("(*"),
+      take_until("*)"),
+      tag("*)")
+    )),
+    |_| () // Output is thrown away.
+  )(i)
+}
+```
+
+## Identifiers
+
+### `Rust-Style Identifiers`
+
+Parsing identifiers that may start with a letter (or underscore) and may contain underscores, letters and numbers may be parsed like this:
+
+```rust
+pub fn identifier(input: &str) -> IResult<&str, &str> {
+  recognize(
+    pair(
+      alt((alpha1, tag("_"))),
+      many0(alt((alphanumeric1, tag("_"))))
+    )
+  )(input)
+}
+```
+
+Let's say we apply this to the identifier `hello_world123abc`.
+The first `alt` parser would recognize `hello_world`. The `pair` combinator ensures that `123abc` will be piped to the next `alphanumeric0` parser, which recognizes every remaining character.
+However, the `pair` combinator returns a tuple of the results of its sub-parsers. The `recognize` parser produces a `&str` of the input text that was parsed.
+
+## Literal Values
+
+### Escaped Strings
+
+This is [one of the examples](https://github.com/Geal/nom/blob/master/examples/string.rs) in the examples directory.
+
+### Integers
+
+The following recipes all return string slices rather than integer values. How to obtain an integer value instead is demonstrated for hexadecimal integers. The others are similar.
+
+The parsers allow the grouping character `_`, which allows one to group the digits by byte, for example: `0xA4_3F_11_28`. If you prefer to exclude the `_` character, the lambda to convert from a string slice to an integer value is slightly simpler. You can also strip the `_` from the string slice that is returned, which is demonstrated in the second hexdecimal number parser.
+
+If you wish to limit the number of digits in a valid integer literal, replace `many1` with `many_m_n` in the recipes: `many1(one_of("_01"))` becomes `many_m_n(1, 16, one_of("_01"))` for binary numbers having at most 16 bits.
+
+#### Hexadecimal
+
+The parser outputs the string slice of the digits without the leading `0x`/`0X`.
+
+```rust
+fn hexadecimal<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str,&'a str, E> {
+  preceded(
+    alt((tag("0x"), tag("0X"))),
+    recognize(
+      many1(
+        one_of("_0123456789abcdefABCDEF")
+      )
+    )
+  )(input)
+}
+```
+
+If you want it to return the integer value instead, use map:
+
+```rust
+fn hexadecimal_value<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, i64, E> {
+  map(preceded(
+    alt((tag("0x"), tag("0X"))),
+      recognize(
+        many1(
+          one_of("_0123456789abcdefABCDEF")
+        )
+      )
+    ),
+    |out: &str| i64::from_str_radix(&str::replace(&out, "_", ""), 16).unwrap()
```

Comment:
`unwrap()` can fail if the literal has too many digits. It would be better to use `map_res()` to handle this case rather than panicking.

Reply:
I debated about how much error handling I should do in these. I think I agree with you, `map_res()` is the right thing to do here.

...but now rustc can't infer the error type.

```
189 |   let x = hexadecimal_value("0xdeadbeef");
    |       -   ^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `E` declared on the function `hexadecimal_value`
    |       |
    |       consider giving `x` the explicit type `std::result::Result<(&str, i64), nom::Err<E>>`, where the type parameter `E` is specified
```

What is the concrete error type in this case? 

Debugging "cannot infer type" errors should be a recipe, actually.

Reply:
I think we should through in a comment here, that the unwrap might fail.
reason:
- The error handling distracts from the goal to communicate just *parser* design
- Rust users are trained to have an alert going on in their heads once they read `unwrap` anyway. They will use this with caution

Reply:
> ...but now rustc can't infer the error type.
> 
> ```
> 189 |   let x = hexadecimal_value("0xdeadbeef");
>     |       -   ^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `E` declared on the function `hexadecimal_value`
>     |       |
>     |       consider giving `x` the explicit type `std::result::Result<(&str, i64), nom::Err<E>>`, where the type parameter `E` is specified
> ```

This doesn't have anything to do with `map_res()`; you would see the same behavior with the original version. It's because the parser is generic over error types `E`, so Nom can't tell what error type you want to use. (Among other options, you can choose between `()` (which stores no information about how the parse failed), `(I, ErrorKind)` (which stores the root cause of the parse failure), and `VerboseError<I>`(which tracks the full context of an error, e.g. `char` failed, inside an `alt`, inside a `many0`).
I agree this is a somewhat awkward feature of Nom and probably deserves a recipe. There are two main solutions: specify `E` explicitly when you call the parser (which is more general), or choose a particular type for `E`.
Here is what specifying a type for `E` looks like:
```rust
let x = hexadecimal_value::<VerboseError<_>>("0xdead_beef");
let y = hexadecimal_value::<(_, ErrorKind)>("0xdead_beef");
```
And fixing a particular `E` just requires changing the function signature. You can use the fact that `IResult` defaults to `E = (I, ErrorKind)`:
```rust
fn hexadecimal_value(input: &str) -> IResult<&str, i64>
```

Reply:
> I think we should through in a comment here, that the unwrap might fail.
> reason:
> 
> * The error handling distracts from the goal to communicate just _parser_ design
> * Rust users are trained to have an alert going on in their heads once they read `unwrap` anyway. They will use this with caution

I agree it feels un-Rusty to not handle the error here. If the parser says it parses an `i64`, then it should either successfully parse an `i64`, or it should fail to parse, but it shouldn't panic. I do think it would be valuable to use `map_res()` here. It actually results in less code, and it is a very idiomatic way to build error handling into the parsing. And it doesn't seem like any of the other examples illustrate `map_res()`. Here's the `map_res()` version, with some indentation fixes:
```rust
map_res(
  preceded(
    alt((tag("0x"), tag("0X"))),
    recognize(many1(one_of("_0123456789abcdefABCDEF"))),
  ),
  |out: &str| i64::from_str_radix(&out.replace("_", ""), 16)
)(input)
```

Reply:
@nilsmartel I would normally agree with you. The reason @calebsander changed my mind in this case is because this is exactly the use case for nom's built-in [`map_res()`](https://docs.rs/nom/5.1.2/nom/combinator/fn.map_res.html). I agree that the reader will be alerted to the need for additional error handling by the presence of `unwrap()`, but if we leave it there, we will be sending the reader down a blind alley, so to speak. They will ultimately want to know about `map_res()` anyway. Including it here tips them off to that need without necessarily cluttering the exposition.